File: 卡码/代码随想录/JS/动态规划/05-子序列问题/04-回文/LeetCode_0647.js
//647. 回文子串
//动态规划
const countSubstrings = function(s) {
	const n=s.length;
	let res=0;
	const dp=new Array(n).fill(0).map(item=>new Array(n).fill(false));
	for(let i=n-1;i>=0;i--){
		for(let j=i;j<n;j++){
			if(s[i]==s[j]&&(j-i<=1||dp[i+1][j-1])){
				dp[i][j]=true;
				res++;
			}
		}
	}
	return res;
};

//双指针
const countSubstrings1 = function(s) {
	let res=0;
	for(let i=0;i<s.length;i++){
		res+=countPalindrome(s,i,i);
		res+=countPalindrome(s,i,i+1);
	}
	return res;
};
const countPalindrome=function(s,i,j){
	let res=0;
	while(i>=0&&j<s.length&&s[i]==s[j]){
		i--;
		j++;
		res++;
	}
	return res;
}

File: 卡码/代码随想录/Java/codeThoughts/src/main/java/com/kurumi/leetCode/codeThoughts/动态规划/_05_子序列问题/_04_回文/LeetCode_0647.java
package com.kurumi.leetCode.codeThoughts.动态规划._05_子序列问题._04_回文;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 647. 回文子串
 */
public class LeetCode_0647 {
    static class Solution {
        // 动态规划
        public int countSubstrings(String s) {
            int n = s.length();
            boolean[][] dp = new boolean[n][n];
            int res = 0;
            for (int i = n - 1; i >= 0; i--) {
                for (int j = i; j < n; j++) {
                    if (s.charAt(i) == s.charAt(j) && (j - i <= 1 || dp[i + 1][j - 1] == true)) {
                        dp[i][j] = true;
                        res++;
                    }
                }
            }
            return res;
        }

        // 双指针
        public int countSubstrings1(String s) {
            int res = 0;
            for (int i = 0; i < s.length(); i++) {
                res += countPalindrome(s, i, i);
                res += countPalindrome(s, i, i + 1);
            }
            return res;
        }

        public int countPalindrome(String s, int i, int j) {
            int res = 0;
            while (i >= 0 && j < s.length() && s.charAt(i) == s.charAt(j)) {
                i--;
                j++;
                res++;
            }
            return res;
        }
    }
}
